File: lib-database-entities/src/main/java/com/ustadmobile/lib/db/entities/SaleProductGroup.java
package com.ustadmobile.lib.db.entities;

import com.ustadmobile.lib.database.annotation.UmEntity;
import com.ustadmobile.lib.database.annotation.UmPrimaryKey;
import com.ustadmobile.lib.database.annotation.UmSyncLastChangedBy;
import com.ustadmobile.lib.database.annotation.UmSyncLocalChangeSeqNum;
import com.ustadmobile.lib.database.annotation.UmSyncMasterChangeSeqNum;

@UmEntity(tableId = 69)
public class SaleProductGroup {

    public static final int PRODUCT_GROUP_TYPE_CATEGORY = 1;
    public static final int PRODUCT_GROUP_TYPE_COLLECTION = 2;
    public static final int PRODUCT_GROUP_TYPE_PRODUCT = 4;

    @UmPrimaryKey(autoGenerateSyncable = true)
    private long saleProductGroupUid;

    //Name of collection. eg: Eid Collection or Toys
    private String saleProductGroupName;

    //Desc eg: This if for eid holidays
    private String saleProductGroupDesc;

    //If collection is to be displayed from one start date
    private long saleProductGroupStartDate;

    //If collection is to be displayed from one start date to one end date
    private long saleProductGroupEndDate;

    //Creation of this collection
    private long saleProductGroupCreationDate;

    private boolean saleProductGroupActive;

    private int saleProductGroupType;

    @UmSyncMasterChangeSeqNum
    private long saleProductGroupMCSN;

    @UmSyncLocalChangeSeqNum
    private long saleProductGroupLCSN;

    @UmSyncLastChangedBy
    private int saleProductGroupLCB;


    public SaleProductGroup(){
        this.saleProductGroupActive = true;
        this.saleProductGroupCreationDate = System.currentTimeMillis();
        this.saleProductGroupType  = PRODUCT_GROUP_TYPE_CATEGORY;
        this.saleProductGroupDesc = "";
    }
    public SaleProductGroup(String name){
        this.saleProductGroupActive = true;
        this.saleProductGroupCreationDate = System.currentTimeMillis();
        this.saleProductGroupType  = PRODUCT_GROUP_TYPE_CATEGORY;
        this.saleProductGroupName = name;
        this.saleProductGroupDesc = "";
    }


    public long getSaleProductGroupUid() {
        return saleProductGroupUid;
    }

    public void setSaleProductGroupUid(long saleProductGroupUid) {
        this.saleProductGroupUid = saleProductGroupUid;
    }

    public String getSaleProductGroupName() {
        return saleProductGroupName;
    }

    public void setSaleProductGroupName(String saleProductGroupName) {
        this.saleProductGroupName = saleProductGroupName;
    }

    public long getSaleProductGroupStartDate() {
        return saleProductGroupStartDate;
    }

    public void setSaleProductGroupStartDate(long saleProductGroupStartDate) {
        this.saleProductGroupStartDate = saleProductGroupStartDate;
    }

    public long getSaleProductGroupEndDate() {
        return saleProductGroupEndDate;
    }

    public void setSaleProductGroupEndDate(long saleProductGroupEndDate) {
        this.saleProductGroupEndDate = saleProductGroupEndDate;
    }

    public long getSaleProductGroupCreationDate() {
        return saleProductGroupCreationDate;
    }

    public void setSaleProductGroupCreationDate(long saleProductGroupCreationDate) {
        this.saleProductGroupCreationDate = saleProductGroupCreationDate;
    }

    public boolean isSaleProductGroupActive() {
        return saleProductGroupActive;
    }

    public void setSaleProductGroupActive(boolean saleProductGroupActive) {
        this.saleProductGroupActive = saleProductGroupActive;
    }

    public int getSaleProductGroupType() {
        return saleProductGroupType;
    }

    public void setSaleProductGroupType(int saleProductGroupType) {
        this.saleProductGroupType = saleProductGroupType;
    }

    public long getSaleProductGroupMCSN() {
        return saleProductGroupMCSN;
    }

    public void setSaleProductGroupMCSN(long saleProductGroupMCSN) {
        this.saleProductGroupMCSN = saleProductGroupMCSN;
    }

    public long getSaleProductGroupLCSN() {
        return saleProductGroupLCSN;
    }

    public void setSaleProductGroupLCSN(long saleProductGroupLCSN) {
        this.saleProductGroupLCSN = saleProductGroupLCSN;
    }

    public int getSaleProductGroupLCB() {
        return saleProductGroupLCB;
    }

    public void setSaleProductGroupLCB(int saleProductGroupLCB) {
        this.saleProductGroupLCB = saleProductGroupLCB;
    }

    public String getSaleProductGroupDesc() {
        return saleProductGroupDesc;
    }

    public void setSaleProductGroupDesc(String saleProductGroupDesc) {
        this.saleProductGroupDesc = saleProductGroupDesc;
    }
}

File: lib-database-entities/src/commonMain/kotlin/com/ustadmobile/lib/db/entities/SaleProductGroup.kt
package com.ustadmobile.lib.db.entities

import androidx.room.Entity
import androidx.room.PrimaryKey
import com.ustadmobile.door.annotation.LastChangedBy
import com.ustadmobile.door.annotation.LocalChangeSeqNum
import com.ustadmobile.door.annotation.MasterChangeSeqNum
import com.ustadmobile.door.annotation.SyncableEntity
import kotlinx.serialization.Serializable


@SyncableEntity(tableId = 69)
@Entity
@Serializable
class SaleProductGroup() {

    @PrimaryKey(autoGenerate = true)
    var saleProductGroupUid: Long = 0

    //Name of collection. eg: Eid Collection or Toys
    var saleProductGroupName: String? = null

    //Desc eg: This if for eid holidays
    var saleProductGroupDesc: String? = null

    //If collection is to be displayed from one start date
    var saleProductGroupStartDate: Long = 0

    //If collection is to be displayed from one start date to one end date
    var saleProductGroupEndDate: Long = 0

    //Creation of this collection
    var saleProductGroupCreationDate: Long = 0

    var saleProductGroupActive: Boolean = false

    var saleProductGroupType: Int = 0

    @MasterChangeSeqNum
    var saleProductGroupMCSN: Long = 0

    @LocalChangeSeqNum
    var saleProductGroupLCSN: Long = 0

    @LastChangedBy
    var saleProductGroupLCB: Int = 0


    init {
        this.saleProductGroupActive = true
        this.saleProductGroupCreationDate = 0
        this.saleProductGroupType = PRODUCT_GROUP_TYPE_CATEGORY
        this.saleProductGroupDesc = ""
    }

    constructor(name: String) : this() {
        this.saleProductGroupActive = true
        this.saleProductGroupCreationDate = 0
        this.saleProductGroupType = PRODUCT_GROUP_TYPE_CATEGORY
        this.saleProductGroupName = name
        this.saleProductGroupDesc = ""
    }

    companion object {

        const val PRODUCT_GROUP_TYPE_CATEGORY = 1
        const val PRODUCT_GROUP_TYPE_COLLECTION = 2
        const val PRODUCT_GROUP_TYPE_PRODUCT = 4
    }
}
